AStar add

mitchellri opened the pull request:
Not sure if you want this included in the project, but it's a simple addition that can be very useful
- Added extensions for AStar
- Added additional comparison and private variable in NodeComparer for heuristic
- Duplicated Dijkstra to AStar file
   - Added heuristic parameter
   - Changed NodeComparer to include heuristic

mitchellri committed:
AStar add

matiii commented:
What are pros of AStar method ?
Why we need heuristic method ?

mitchellri commented:
AStar can increase performance of the pathfinding algorithm ([Video example](https://youtu.be/g024lzsknDo))
Heuristic is a cost function that helps determine which node to evaluate next ([Stackoverflow](https://stackoverflow.com/a/43567824/4725050))

_Note: Since I made this request I have made additional commits to my master branch. If you wish to add AStar to your master, make sure to either reject the unrelated changes if possible, I can submit a new pull request omitting the unrelated commits, or change my master branch to only include the AStar add commit_

matiii commented:
Hi, really nice feature. Could you provide more refinement to your contribution ? I prefer to keep graph data stracture still immutable. Instead modify NodeComparer would be better just extend it by inheritance. Why we need to expose herusitc method to be public ? Is it posible to have default implementation ?

mitchellri committed:
Refactor heuristic node comparing

mitchellri commented:
Thanks! Dijkstra's algorithm is a special case of the AStar algorithm when it has null heuristic, so **the default implementation of AStar is Dijkstra's algorithm**. 
- Removed commits up to original AStar implementation (Graph remains mutable)
- NodeComparer original implementation with modified access for child classes
- Added HeuristicNodeComparer extending NodeComparer to include heuristic

In this implementation AStar requires a heuristic, because Dijkstra should be used for Dijkstra.

Optionally, I could allow null heuristic in AStar which would perform Dijkstra's algorithm. In that case it might be worth considering removing Dijkstra if you do not want redundancy, but it might be confusing to have no Dijkstra in Dijkstra.NET
Here's what that might look like:
```c#
public static ShortestPathResult GetShortestPath(IDijkstraGraph graph, uint from, uint to, int depth, Func<uint, uint, int> heuristic = null)
{
        var path = new Dictionary<uint, uint>();
        var distance = new Dictionary<uint, int> {[from] = 0};
        var d = new Dictionary<uint, int> {[from] = 0};
        var comparer = heuristic is null ? new NodeComparer(distance) : new HeuristicNodeComparer(distance, heuristic);
        var q = new SortedSet<uint>(new[] {from}, comparer);
        var current = new HashSet<uint>();
        ...
```

matiii commented:
Now, it looks much better. Could you provide unit tests as well ? Then we will be certain about reliabilty your solution.
